plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/contributors/FirCallableCompletionContributor.kt

```diff
@@ -167,11 +184,79 @@ internal open class FirCallableCompletionContributor(
         val typeOfPossibleReceiver = explicitReceiver.getKtType() ?: return
         val possibleReceiverScope = typeOfPossibleReceiver.getTypeScope() ?: return
 
+        if (explicitReceiver is KtSuperExpression) {
+            collectDotCompletionForSuperAsReceiver(
+                explicitReceiver,
+                typeOfPossibleReceiver,
+                visibilityChecker,
+                expectedType
+            )
+            return
+        }
+
         val nonExtensionMembers = collectNonExtensions(possibleReceiverScope, visibilityChecker, scopeNameFilter) { filter(it) }
-        val extensionNonMembers = collectSuitableExtensions(implicitScopes, extensionChecker, visibilityChecker)
+        collectNonExtensionLookUpElements(nonExtensionMembers.toList(), expectedType)
+
+        val extensionMembers = collectSuitableExtensions(implicitScopes, extensionChecker, visibilityChecker)
+        extensionMembers.forEach { (symbol, applicability) ->
+            getExtensionOptions(symbol, applicability)?.let { addCallableSymbolToCompletion(expectedType, symbol, it, getPriority(symbol)) }
+        }
+
+        collectTopLevelExtensionsFromIndices(listOf(typeOfPossibleReceiver), extensionChecker, visibilityChecker)
+            .filter { filter(it) }
+            .forEach { addCallableSymbolToCompletion(expectedType, it, getOptions(it) ?: return@forEach, getPriority(it)) }
+    }
+
+    protected fun KtAnalysisSession.collectDotCompletionForSuperAsReceiver(
+        superExpression: KtSuperExpression,
+        superExpressionType: KtType,
+        visibilityChecker: CompletionVisibilityChecker,
+        expectedType: KtType?,
+    ) {
+        if (superExpressionType !is KtIntersectionType) {
+            val nonExtensionMembers = getNonExtensionsMemberSymbols(superExpressionType, visibilityChecker)
+            collectNonExtensionLookUpElements(nonExtensionMembers.toList(), expectedType)
+            return
+        }
+        val superTypes = superExpressionType.conjuncts
+        val allSymbols = mutableListOf<KtCallableSymbol>()
+        val symbolsByName = mutableMapOf<Name, AtomicInteger>()
+        for (superType in superTypes) {
+            for (symbol in getNonExtensionsMemberSymbols(superType, visibilityChecker)) {
+                symbolsByName.computeIfAbsent(symbol.callableIdIfNonLocal?.callableName ?: continue) { AtomicInteger() }
+                    .incrementAndGet()
+                allSymbols.add(symbol)
+            }
+        }
+
+        val nameNeedDisambiguation = mutableSetOf<Name>()
+        for ((name, count) in symbolsByName) {
+            if (count.get() > 1) {
+                nameNeedDisambiguation.add(name)
+            }
+        }
```

Comment:
All that this piece of code does is collects `nameNeedDisambiguation`. It should be extracted to function

Reply:
Sounds good! Thanks!